framework/src/org/checkerframework/common/value/Range.java

```diff
@@ -0,0 +1,433 @@
+package org.checkerframework.common.value;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+
+/**
+ * The Range Class with mathematics operations. Models the range indicated by the @IntRange
+ * annotation.
+ *
+ * <p>Note that this class is not responsible for detecting incorrect parameters, e.g. the value of
+ * "from" could be greater than the value of "to". This incorrectness would be eventually caught by
+ * {@link
+ * org.checkerframework.common.value.ValueAnnotatedTypeFactory#createIntRangeAnnotation(Range)} when
+ * creating an annotation from range, which would then be replaced with @UnknownVal.
+ *
+ * @author JasonMrX
+ */
+public class Range {
+
+    /** The value 'from' */
+    public final long from;
+
+    /** The value 'to' */
+    public final long to;
+
+    /**
+     * Constructs a range with its bounds specified by two parameters, "from" and "to".
+     *
+     * <p>Note that it is possible to construct a range with incorrect parameters, e.g. the value of
+     * "from" could be greater than the value to "to". This incorrectness would be caught by {@link
+     * org.checkerframework.common.value.ValueAnnotatedTypeFactory#createIntRangeAnnotation(Range)}
```

Comment:
I strongly feel that `Range` should be treated as an abstraction in its own right and should be understandable by itself.  It should be decoupled from how it happens to be used by the Value Checker.  It should never have an illegal representation.  Making these changes will make the `Range` code cleaner and clearer.  It will also fix bugs.  For example, when an intersection is empty, the corresponding `@IntRange` annotation needs to be bottom, not top as it currently is.  I would add a special `NOTHING` constant to `Range` and have operations check it first.  When ValueAnnotatedTypeFactory sees `NOTHING`, the corresponding `@IntRange` annotation can be bottom.